### PR TITLE
Improved the filterBy function with OR condition and custom filter algorythm

### PR DESCRIPTION
**Pull request by Falseee**
This PR improves the filterBy function, its basicly adds a or condition check(requested by @thomas-nguyen-3  in #3295) and additionally i added a option to use a custom filter algorithm.

The new functionalities are documented.

Jsfiddle with all 3 filter algorithm: https://jsfiddle.net/j1o6p4s8/
In the Jsfiddle the custom algorithm will remove all rows which match with at least one filter!

### CODE REVIEW

File: src/bootstrap-table.js [hunk 1]
```diff
@@ -1353,19 +1356,43 @@
...
-              (Array.isArray(f[key]) &&
-              !f[key].includes(item[key])) ||
-              (!Array.isArray(f[key]) &&
-              item[key] !== f[key])
-            ) {
-              return false
+        if (typeof this.filterOptions.filterAlgorithm === 'function') {
+          this.data = this.options.data.filter((item, i) => {
+            return this.filterOptions.filterAlgorithm.apply(null, [item, f]);
+          });
```
Comment (wenzhixin) [inline src/bootstrap-table.js]: need to fix the eslint problem.

File: src/bootstrap-table.js [hunk 2]
```diff
@@ -1569,10 +1596,6 @@
           to = o.paginationSuccessivelySize
         }
 
-        if (to > this.totalPages) {
-          to = this.totalPages
-        }
-
```
Comment (wenzhixin) [inline src/bootstrap-table.js]: deleting these columns?

**Comment by djhvscf** [comment]
Any update @Falseee

File: src/bootstrap-table.js (hunk 2 above)
Comment (Falseee) [reply src/bootstrap-table.js]: Yes!
Check https://github.com/wenzhixin/bootstrap-table/blob/develop/src/bootstrap-table.js#L1584-L1586 and https://github.com/wenzhixin/bootstrap-table/blob/develop/src/bootstrap-table.js#L1596-L1598 code duplication.

**Comment by Falseee** [comment]
@djhvscf i will update it today.

File: src/bootstrap-table.js (hunk 1 above)
Comment (Falseee) [reply src/bootstrap-table.js]: done

**Comment by charliehewis** [comment]
This has been massively helpful in working out how to use the custom filter algorithm, thankyou

**Comment by UtechtDustin** [comment]
@charliehewis It is doumentated (https://bootstrap-table.com/docs/api/methods/#filterby), in my post above is als a jsfiddle which contains examples.
You just have to use/set the `filterAlgorithm` option.

**Comment by charliehewis** [comment]
@UtechtDustin The documentation is not very helpful with respect to the custom filterAlgorithm and the example given there doesn't use it. I've spent a lot of time over the past few weeks with filters trying to get it to work and the example you gave here was the key to understanding it. I've also struggled with the persistence of the filter, clearing it and a few other things so I'd be happy to write some extra docs or examples that would help other people in my position

**Comment by UtechtDustin** [comment]
@charliehewis Okay, could you please explain what do you try to achieve ?
Also please provide us an example (using our [editor](https://live.bootstrap-table.com/)) what do you already tried.

**Comment by charliehewis** [comment]
@UtechtDustin I used the editor to make this example but I don't know where the save button on the editor is actually saving it to, or how to give you a link to what I wrote. I also can't get the right settings on jsfiddle. I've put it in a text file for now (sorry).

In my project I have Bootstrap tables on a series of webpages. The project is in typescript, using ReactJS but not react-bootstrap. I have found that when setting the filter on a table on one page, that same custom filter is used for a different table on a different page as if it is persisting even when it doesn't make sense to do so (because the data-fields have changed and the id of the new table is different). In order to fix this, I have had to include upon page load a method which updates the filter to clear it by explicitly returning to default $('#table').bootstrapTable('filterBy', {}, { 'filterAlgorithm': 'and'})

[JQuery_Bootstrap_Table_Filter.txt](https://github.com/wenzhixin/bootstrap-table/files/4687853/JQuery_Bootstrap_Table_Filter.txt)

**Comment by UtechtDustin** [comment]
@charliehewis on the top is a save button, after you saved the url has be changed to your example url.

**Comment by charliehewis** [comment]
@UtechtDustin Ah yes I see now, thanks. This is my example which demonstrates how you can use the custom filter with a key that isn't one of the data-fields, but it could be tidied up by having the 'choice' value be passed in as some property of the button clicked so that only one function is needed:
https://live.bootstrap-table.com/code/charliehewis/3169

**Pull request merged**